py/persistentcode: Add architecture flags compatibility checks.

agatti opened the pull request:
### Summary

As part of the work to make MicroPython use selected RISC-V extensions to improve generated code coming from built-in compiler and from `mpy-cross`, `Zba` opcodes support was added in #17635 without the ability to perform checks on its availability on the running environment when loading MPY files making use of those opcodes.

Now, this commit extends the MPY file format in a backwards-compatible way to store an encoded form of architecture-specific flags that have been specified in the "mpy-cross" command line, or that have been explicitly set as part of a native emitter configuration.

Following the suggestion in https://github.com/micropython/micropython/pull/17635#issuecomment-3274894017, the file format changes are as follows:

* The features byte, previously containing the target native architecture and the minor file format version, now claims bit 6 as a flag indicating the presence of an encoded architecture flags integer
* If architecture flags need to be stored, they are placed right after the MPY file header

This means that properly-written MPY parsers, if encountering a MPY file containing encoded architecture flags, should raise an error since no architecture identifiers have been defined that make use of bits 6 and 7 in the referenced header byte.  This should give enough guarantees of backwards compatibility when this feature is used (improper parsers were subjected to breakage anyway).

The meaning of the architecture flags value is backend-specific, with the only common characteristic of being a variable-encoded unsigned integer for the time being.

The changes made to the file format effectively limit the number of possible target architectures to 16, of which 13 are already claimed. There aren't that many new architectures planned to be supported for the lifetime of the current MPY file format, so this change still leaves space for architecture updates if needed.

### Testing

Since #17635 forced all MPY files built by the `VIRT_RV32` board in the QEMU port to have `Zba` opcodes enabled when running tests using `--via-mpy`, said port was used for preliminary testing of pass/reject conditions on MPY files' architecture flags (`Zba` opcodes are enabled by default for QEMU):

* `mpy-cross` had to be rebuilt to pick up the new changes related to MPY file saving
* The full test suite was run using `make BOARD=VIRT_RV32 test_full`, to check that MPY files with extension requirements that match the ones available are accepted
* `MICROPY_EMIT_RV32_ZBA` was disabled in `ports/qemu/mpconfigport.h`, the port was rebuilt, and the full test suite was ran again to make sure that MPY files requiring extensions that are not available are rejected 
* `-march-flags=zba` was removed from `ports/qemu/boards/VIRT_RV32/mpconfigboard.mk`, and the full test suite was ran again to make sure that MPY files without extension requirements would be accepted if no extensions are involved neither in the MPY files nor in the execution environment
* As a final sanity check, `MICROPY_EMIT_RV32_ZBA` was re-enabled in `ports/qemu/mpconfigport.h`, the port was rebuilt, and the full test suite was ran once more to make sure MPY files with no extension requirements are still accepted even if there are other extensions available

### Trade-offs and Alternatives

This will come with a minimal size increase for all port configurations that can load persistent code.  The alternative is to gate part of the checks behind another set of definitions, with the effect of increasing configuration complexity.

`tools/mpy_arch_flags.py` was added as a separate file to enable quick testing of different scenarios.  I've looked into adding its function (create a copy of a MPY file but with updated arch flags) into `tools/mpy-tool.py` but fitting that in the current options model would require too many changes for something that, in theory, shouldn't be needed by most end users.  If it's deemed to be not really needed as part of MicroPython it can be safely taken out of the PR.

dpgeorge commented on py/bc.h:
```diff
@@ -220,6 +220,7 @@ typedef struct _mp_compiled_module_t {
     bool has_native;
     size_t n_qstr;
     size_t n_obj;
+    size_t arch_flags;
```
Can we combine the storage with `has_native`?  Eg:
```
    bool has_native : 1;
    uint32_t arch_flags : 31;
```

agatti replied:
In theory yes, it can be done, at the cost of propagating this "uncommon" 31-bits limit throughout the MPY toolchain.

However, from what I can see, all instances of `struct mp_compiled_module_t` in MicroPython's source tree are transient.  I'm not really sure if it is really needed to increase complexity down the line for this right now, though.

Assuming stack space reduction is critical for this particular scenario, well, then probably `n_qstr`'s type in that struct should be replaced with a `uint16_t`.  If there's a hard limit on the number of QSTRs in the pool (otherwise all usages of `qstr_short_t` will break due to index wraparound) then it cannot be possible to have a working interpreter after loading a MPY file with more than 64K strings anyway, so it using a `size_t` isn't helping there.

If the compiler is smart enough, rewriting the struct as

```c
typedef struct _mp_compiled_module_t {
    mp_module_context_t *context;
    const struct _mp_raw_code_t *rc;
    #if MICROPY_PERSISTENT_CODE_SAVE
    size_t n_obj;
    size_t arch_flags;
    uint16_t n_qstr;
    bool has_native;
    #endif
} mp_compiled_module_t;
```

should both keep the arch flags with a word-aligned size and save a byte.  If you also limit `arch_flags` to a `uint8_t` for now and then place it at the end of the struct then you'll save four bytes instead.  Maybe `n_obj` could receive the same narrowing treatment, I haven't yet looked into object deserialisation.

This also led me to an interesting rabbit hole.  `read_uint` in `py/persistentcode.c` doesn't have a stop condition so an appropriately malformed file may end up reading the wrong data, possibly crashing the board or worse.  In theory devices could be compromised that way if, with a appropriatedly manipulated vuint loads, the original MPY natmod trampoline is not executed but the loader jumps somewhere else in the code segment.

I'm exaggerating on that, I know, and I'm sure the security implications have already been considered before.

Still, if hardening vuint loading and patching `mpy-tool.py` to reject files that would trigger a QSTR index wraparound is something worth adding I can tack those two commits to this PR when I split the changes into separate commits, or just open a new PR :)

dpgeorge commented on py/persistentcode.c:
```diff
@@ -471,7 +475,7 @@ void mp_raw_code_load(mp_reader_t *reader, mp_compiled_module_t *cm) {
         mp_raise_ValueError(MP_ERROR_TEXT("incompatible .mpy file"));
     }
-    if (MPY_FEATURE_DECODE_ARCH(header[2]) != MP_NATIVE_ARCH_NONE) {
+    if (arch != MP_NATIVE_ARCH_NONE) {
```
It looks like the check now ignores the MSB of `arch`.  It would be good to reserve that bit for future use, and so force it to be 0 here.

An easy way to do that would be:
```c
#define MPY_FEATURE_DECODE_ARCH(feat) (((feat) >> 2) & 0x2F)
```

dpgeorge commented on py/persistentcode.c:
```diff
@@ -483,6 +487,26 @@ void mp_raw_code_load(mp_reader_t *reader, mp_compiled_module_t *cm) {
 
+    size_t arch_flags = 0;
+    if (MPY_FEATURE_ARCH_FLAGS_TEST(header[2])) {
+        // TODO: Report an error for targets that do not have support for
```
I suggest that if the flags bit is set, it should always do the `read_uint`.  That makes the parser more robust.

Then, for non-RV32 archs, those flags should be 0.

agatti replied:
Originally the read would have been done in all cases, but having an unexpected flag bit is a signal in itself that something's not right with the file and thus it's not right to proceed.

I moved that behind a platform ifdef to reduce the footprint increase, since this affects all ports.  You can check the history of https://github.com/micropython/micropython/pull/18106#issuecomment-3315085609 to see the impact on non-RV32 ports.

dpgeorge commented on py/persistentcode.c:
```diff
@@ -672,15 +697,34 @@ void mp_raw_code_save(mp_compiled_module_t *cm, mp_print_t *print) {
+
     mp_print_bytes(print, header, sizeof(header));
 
+    // The encoded architecture flags could have been placed at the end, but:
```
I think this justification can be removed from here.  It can go in the commit message.

agatti committed:
py/persistentcode: Add architecture flags compatibility checks.

This commit extends the MPY file format in a backwards-compatible way to
store an encoded form of architecture-specific flags that have been
specified in the "mpy-cross" command line, or that have been explicitly
set as part of a native emitter configuration.

The file format changes are as follows:

* The features byte, previously containing the target native
  architecture and the minor file format version, now claims bit 6 as a
  flag indicating the presence of an encoded architecture flags integer
* If architecture flags need to be stored, they are placed right after
  the MPY file header.

This means that properly-written MPY parsers, if encountering a MPY file
containing encoded architecture flags, should raise an error since no
architecture identifiers have been defined that make use of bits 6 and
7 in the referenced header byte.  This should give enough guarantees of
backwards compatibility when this feature is used (improper parsers were
subjected to breakage anyway).

The encoded architecture flags could have been placed at the end, but:

* Having them right after the header makes the architecture
  compatibility checks occur before having read the whole file in memory
  (which still happens on certain platforms as the reader may be backed
  by a memory buffer), and prevents eventual memory allocations that do
  not take place if the module is rejected early
* Properly-written MPY file parsers should have checked the upper two
  bits of the flags byte to be actually zero according to the format
  specification available right before this change, so no assumptions
  should have been made on the exact order of the chunks for an
  unexpected format.

The meaning of the architecture flags value is backend-specific, with
the only common characteristic of being a variable-encoded unsigned
integer for the time being.

The changes made to the file format effectively limit the number of
possible target architectures to 16, of which 13 are already claimed.
There aren't that many new architectures planned to be supported for the
lifetime of the current MPY file format, so this change still leaves
space for architecture updates if needed.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tools/mpy-tool.py: Add architecture flags to disassembly output.

This commit extends "mpy-tool.py"'s disassembly output of a given MPY
file (triggered via the "-d" command line option) to include newly added
fields.

Now the target architecture for the chosen MPY file is printed out to
screen in human-readable format and, if present, architecture-specific
flags.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/persistentcode: Add architecture flags check for RV32 platforms.

This commit introduces the MPY architecture flags checking code specific
for the RV32 target, currently checking for the only additional
extension that is supported by the runtime: Zba.

The warnings inside "mpy-cross" have also been removed since now there
is a way to reject incompatible MPY files at runtime.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/modsys: Add architecture flags to MicroPython metadata.

This commit adds the currently supported architecture flags value as the
upper part of "sys.implementation._mpy".

This had the side effect of perturbing quite a bit of testing
infrastructure and invalidating documentation related to MPY files.  To
make the test suite run successfully and keep the documentation in sync
the following changes have been made:

* The target info feature check file now isolates eventual architecture
  flags and adds them as a separate field
* The test runner now picks up the new architecture flags field, reports
  it to STDOUT if needed and stores it for future uses
* Relevant test files for MPY files import code had to be updated to
  mask out the architecture flags bits in order to perform correctly
* MPY file format documentation was updated to show how to mask off and
  properly display the architecture flags information.

This works out of the box if the flag bits can fit in a smallint value
once merged with the MPY file header value.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tests/micropython: Extend MPY import tests for RV32.

This commit extends the "micropython/import_mpy_native_gc" test with a
pre-built serialised MPY file built for RV32.

Before this commit, the test was skipped on said platform due to the
lack of a known-good MPY file to test the import procedure against.
Now the test is executed as part of the general CI test checks.

The MPY file's Makefile script was also updated to cater for modern
Linux environments where the required Python interpreter's command name
isn't "python", but "python3" instead (this occurs when using the past
two LTS Ubuntu releases for example).  This is the same kind of change
made in 31a008c6e26eccc3798a9ab4169019a02eadb830.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tests/run-tests.py: Pass auto-detected architecture flags to mpy-cross.

This commit lets "run-tests.py" use the encoded architecture flags
provided by the interpreter when invoking "mpy-cross".

If architecture flags are detected, they're mapped into the necessary
strings needed by "mpy-cross"'s "-march-flags" argument, so that tests
will always use all available extensions reported by the target.

Currently this is limited to the RV32 platform, as it is the only one
that is making use of this facility as of now.  This also lets the QEMU
port remove forced arguments to "mpy-cross" when running the test suite.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tests/ports/qemu: Add arch flags MPY import checks test.

This commit introduces a test file (along with its necessary data) to
test MPY files pass/reject cases, due to either mismatching architecture
flag requirements or unneeded "no-flags" values set.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
The PR now should cover all the requirements:

* Core checks taking not too much space on all ports that can load code at runtime
* Updated documentation to be in sync with the rest of the changes
* `mpy-tool.py` is able to manipulate architecture flags to a certain extent
* Automated tests for QEMU
* Hardware test runs (see Octoprobe runs 344 and 345)
* Test runner now autodetect architecture flags and automatically passes them to `mpy-cross` for `--via-mpy` and `--via-mpy --emit native` tests.

I hope I didn't miss anything in the list above!

dpgeorge commented on tests/ports/qemu/mpy_arch_flags_test.py:
```diff
@@ -0,0 +1,128 @@
+
+    sys.implementation._mpy
+    io.IOBase
+except ImportError:
```
This will need to be `(ImportError, AttributeError)` to catch the case of `._mpy` or `io.IOBase` not existing.

dpgeorge commented on docs/reference/mpyfiles.rst:
```diff
@@ -153,10 +159,42 @@ size    field
+MPY files targeting RV32 or RV64 that do not need any particular processor
+extensions do not need to provide a flags integer (along with setting the
+appropriate bit in the header).  The lack of a flags value for RV32 and RV64
+MPY files is used to indicate that no specific extensions are needed, saves one
```
*and* saves one byte ...

dpgeorge commented on tests/ports/qemu/mpy_arch_flags_test_module/Makefile:
```diff
@@ -0,0 +1,13 @@
+
+.PHONY: all
+all:
+	@$(MPY_CROSS) -march=$(ARCH) $(SRC) -o - | $(PYTHON) -c 'import sys; print(sys.stdin.buffer.read())'
```
I don't think `-march=` is needed here, because there's no native code in `module.py`.

And maybe this makefile is not needed at all if you just want to write in the test a comment like:
```py
# To recreate this string run:
# echo "hello=1" | mpy-cross -s module.py - | python -c 'import sys; print(sys.stdin.buffer.read())'
```

dpgeorge commented on tests/ports/qemu/mpy_arch_flags_test_module/module.py:
```diff
@@ -0,0 +1,2 @@
+def hello():
+    print("hello")
```
This whole file could just be `hello = 1` if you want to simplify things (up to you).

Also see above comment about possibly removing this file altogether.

dpgeorge reviewed (commented):
Thanks for updating this PR, and adding docs and tests.  It's looking really good now, and I think the addition of `arch_flags` to `sys.implementation._mpy` makes sense.

Just a few minor details/comments to address then this will be good to merge.

agatti committed:
mpy-cross/main: Parse raw integer arch flags values too.

This commit extends "mpy-cross"'s parsing of the architecture flags
value command line, allowing raw integer values as well as flag strings.

Integers can be represented as either decimal, binary, or hexadecimal
values, using the usual C language prefixes to mark non-base 10 values.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented on docs/reference/mpyfiles.rst:
```diff
@@ -58,12 +58,14 @@ If importing an .mpy file fails then try the following:
     if arch:
         print(' -march=' + arch, end='')
+    if (sys_mpy >> 16) != 0:
+        print(' -march-flags=' + (sys_mpy >> 16), end='')
```
That's a good point.  I've added integer flags support to `mpy-cross` leveraging `mp_parse_num_integer` to do the actual parsing.  This should handle the output of the documentation code snippet.

It can currently handle decimal, binary, and hexadecimal digits.  Probably binary values aren't really needed but given that we're talking about raw bitfields here then it may be easier to represent values that way too.

In theory `tests/run-tests.py` could be modified to use raw integer values to rather than passing a set of human-readable flag names to `mpy-cross`.  It doesn't yield any simplifications since the mapped strings are to be printed to STDOUT anyway, but it's something that can be done if needed.

dpgeorge replied:
OK, what you've done here is good now.

agatti commented:
The PR should be ready now - I've reworked a few more things w.r.t. comments and commit messages, but it should be stable now.

Also, the footprint on rp2 was halved: the redundant stray flag bits check wasn't taken out by the compiler.

dpgeorge commented on tests/ports/qemu/mpy_arch_flags_test.py:
```diff
@@ -0,0 +1,128 @@
+    while val != 0:
+        encoded.append(0x80 | (val & 0x7F))
+        val >>= 7
+    return bytes([encoded[i] for i in range(len(encoded) - 1, -1, -1)])
```
Oh, you're right, MicroPython doesn't have `bytearray.insert`.



> If `bytearray.insert` is indeed not available then `tools/mpy-tool.py` may need to be updated to run with MicroPython too, however that's definitely out of scope of this PR.

Hmm, you're right that `mpy-toll.py` is intended to be able to be run with MicroPython, see e1b2f2e078727999148a71ab969c1685173d25c0 .  And there's a CI test for that which passes... that's because the CI test does not test merging of mpy files which is the part of `mpy-tool.py` that uses `bytearray.insert`.

I guess that's a PR for another day.

dpgeorge reviewed (approved):
Thanks for updating.  And thanks for splitting the changes into nice, clean commits with commit messages that have just the right amount of detail.  Also, great to have corresponding docs and tests.  I know that it's a lot of work to cross all the t's and dot all the i's for this kind of change.
